src/shared/ui/pickerPrompter.ts

```diff
@@ -274,30 +301,49 @@ export class QuickPickPrompter<T> extends Prompter<T> {
      * previously selected item will remain selected if it still exists after loading.
      *
      * @param items DataQuickPickItems or a promise for said items
+     * @param disableInput Disables the prompter until the items have been loaded, only relevant for async loads (default: true)
      * @returns A promise that is resolved when loading has finished
      */
-    public loadItems(items: Promise<DataQuickPickItem<T>[]> | DataQuickPickItem<T>[]): Promise<void> {
+    public async loadItems(items: ItemLoadTypes<T>, disableInput: boolean = true): Promise<void> {
+        // This code block assumes that callers never try to load items in parallel
+        // For now this okay since we don't have any pickers that require that capability
         const picker = this.quickPick
 
-        if (this.hasPlaceholder) {
+        if (this.isShowingPlaceholder) {
             this.clearItems()
         }
 
-        if (items instanceof Promise) {
-            picker.busy = true
-            picker.enabled = false
-            this.pendingPromises += 1
-
-            return items.then(items => {
-                this.pendingPromises -= 1
-                picker.busy = this.pendingPromises !== 0
-                this.appendItems(items)
-                picker.enabled = true
-            })
+        picker.busy = true
+        picker.enabled = !disableInput
+
+        if (isAsyncIterable(items)) {
+            // Technically AsyncIterators have three types: one for yield, one for return, and one
+            // for parameters to `next`. We only care about the first two, where the yield type will
+            // always be the same as the AsyncIterable type variable, and the second will potentially
+            // be undefined
+            const iterator = items[Symbol.asyncIterator]() as AsyncIterator<
+                DataQuickPickItem<T>[],
+                DataQuickPickItem<T>[] | undefined
+            >
+            while (true) {
```

Comment:
**STRONGLY** consider a way to halt iteration if an option is chosen while loading. I'm assuming this is a replacement for the [IteratingQuickPickController](https://github.com/aws/aws-toolkit-vscode/blob/85f5ecac9d4abfcbcaa83c0d0ee93ec7ad935d62/src/shared/ui/picker.ts#L170), which specifically handled pausing and manual pagination.

Reply:
Technically you can use `IteratingQuickPickController` with this prompt class, though ideally the new class should be able to cover the same functionality. The main thing is just making sure the controller object is not created when the prompt is because prompts are meant to be 'stateless' (at least from the wizard's perspective). For now there won't be a 'pause' while loading, but it will get implemented when the need arrives.